libs/llrb/llrb.go

```diff
@@ -0,0 +1,256 @@
+package llrb
+
+import (
+	"fmt"
+	"math"
+	"sync"
+)
+
+// MaxSize is the max allowed number of node a llrb is allowed to contain
+const MaxSize = int(^uint(0) >> 1)
+
+func (a NodeKey) compare(b NodeKey) int {
+	if a.Priority > b.Priority {
+		return 1
+	}
+	if a.Priority < b.Priority {
+		return -1
+	}
+	if a.TS.Before(b.TS) {
+		return 1
+	}
+	if a.TS.After(b.TS) {
+		return -1
+	}
+	return 0
+}
+
+type node struct {
+	key         NodeKey
+	data        interface{}
+	left, right *node
+	black       bool
+}
+
+type llrb struct {
+	mtx     sync.RWMutex
+	size    int
+	maxSize int
+	root    *node
+}
+
+// newLLRB return llrb with given maxSize
+func newLLRB(maxSize int) *llrb {
+	return &llrb{maxSize: maxSize}
+}
+
+// Size returns the number of nodes in the tree
+func (t *llrb) Size() int {
+	t.mtx.RLock()
+	defer t.mtx.RUnlock()
+	return t.size
+}
+
+// GetNext retrieves a satisfied tx with "largest" nodeKey and "smaller" than starter if provided
+func (t *llrb) GetNext(starter *NodeKey, predicate func(interface{}) bool) (interface{}, error) {
+	t.mtx.RLock()
+	defer t.mtx.RUnlock()
+	startKey := NodeKey{
+		Priority: uint64(math.MaxUint64),
+	}
+	if starter != nil {
+		startKey = *starter
+	}
+	var candidate *node
+	for h := t.root; h != nil; {
+		if h.key.compare(startKey) == -1 && predicate(h.data) {
```

Comment:
I suggest `predicate` can be `nil` and should be checked against, then users won’t need to provide a dummy predicate function always returning true.

Reply:
I think what the user should provide is gas and bytes limit, and these two limitations will be packed by llrbmempool automatically, so we can either check func here or make sure that mempool will pass a valid func?

Reply:
```
if h.key.compare(startKey) == -1 && (predicate == nil || predicate(h.data))
```